torch/nn/parallel/distributed.py

```diff
@@ -864,6 +866,49 @@ def __init__(
 
         self._lazy_init_ran = False
 
+        # Register the AccumulateGrad post hooks even if DDP is not compiled. This
+        # can avoid compiling the hooks twice. The Python hooks will be
+        # deregistered later if DDP is not compiled.
+        self._accum_grad_hooks: List[RemovableHandle] = []
+        self._ddp_python_hook = torch._dynamo.config.ddp_python_hook
+        self._force_to_disable_reducer = False
+        if self._ddp_python_hook:
+            if torch._dynamo.config.optimize_ddp:
+                raise RuntimeError(
+                    "Only one of `torch._dynamo.config.optimize_ddp` and "
+                    "`torch._dynamo.config.ddp_python_hook` should be True."
+                )
+            self._register_accum_grad_hook()
+
+    def _register_accum_grad_hook(self):
+        import torch.distributed._functional_collectives as fcol
+
+        def compiled_accum_grad_hook(
+            param,
+            *,
+            param_index: int,
+        ):
+            if not self.require_backward_grad_sync:
+                return
+
+            if self._comm_hooks:
+                for hook, state in self._comm_hooks:
+                    hook(state, (param.grad, param))
+            else:
+                gradient = param.grad / self.process_group.size()
```

Comment:
why do we divide the gradient here?

Reply:
DDP performs the average. We need to match the eager behavior.